rfc/state-model.rst

```diff
@@ -0,0 +1,299 @@
+
+        The Store is ready for a new firmware update to be started.
+
+    * - WRITING
+      - The Client is writing a new firmware image to the *second*, in preparation for installation.
+
+        When writing is complete, it can be installed.
+
+    * - STAGED
+      - Installation of the *second* has been requested, but the system must be restarted as the final update operation runs within the Bootloader.
+
+        This state is transient.
+
+    * - FAILED
+      - An installation of the *second* has been attempted, but has failed for some reason. The failure reason is recorded in the Store.
+
+        The *second* needs to be erased before another update can be attempted.
+
+    * - TRIAL
+      - Installation of the *second* has succeeded, and is now the *active* running in 'trial mode'. This state is transient, and requires the Client to explicitly accept the trial to make the update permanent.
+
+        In this state, the previously installed *active* image is preserved as the *second*. If the trial is explicitly rejected, or the system restarts without accepting the trial, the previously installed image is re-installed and the trial image is rejected.
+
+    * - REJECTED
+      - The *active* trial image has been rejected, but the system must be restarted so the bootloader can revert to the previous image (saved as the *second*).
+
+        This state is transient.
+
+    * - UPDATED
+      - The *active* trial image has been accepted, and is now permanently active.
+
+        The *second* contains the now-expired previous firmware image, which needs to be erased before another update can be started.
+
+Note
+    In the basic flow here, REJECTED seems to have limited difference from TRIAL, other than that it records that the trial has been explicitly rejected.
+
+    However, it provides two important capabilities:
+
+    * It specifically prevents the Client accepting a trial after rejecting it
+    * It enables other TRIAL policies, such as permitting a limited number of restarts (but more than one) before automatically reverting to the previous image.
+
+State transitions
+~~~~~~~~~~~~~~~~~
+
+The state transitions occur either as a result of an API call from the Client, or when the Bootloader carries out an installation operation. The installation operations that occur when the Bootloader runs are determined by the state of the Firmware Store, and do not depend on the reason for the restart. To ensure a specific action occurs at restart, the Client uses the appropriate API to set the state of the Firmware Store.
+
+The Client can trigger transitions in the state model using the following operations:
+
+* ``start``
+* ``write``
+* ``install``
+* ``erase``
+* ``accept``
+* ``reject``
+
+Some transitions can only be triggered by a system restart, which will run the Bootloader. For these transitions it does not matter how the reboot was triggered.
```

Comment:
This is not strictly true for some image types (eg external radio or partial firmware data). Maybe making it less explicit about reset and adding a clarification that's it's not as much about system reset but about use-case specific process of applying the update, would make it more accurate.

Reply:
I should have started the flow graphic with a description of the characteristics of the component to which this flow applies. Reboot being required to complete installation is a assumption for this graphic.

If a component does not require a reboot to apply the update, then a subset of the state model is needed, with a slightly simplified flow. It might look something like this:

![image](https://user-images.githubusercontent.com/5690199/169879864-b5c0c5ed-1e17-4f63-9e5c-2cb9e0e95197.png)

In this suggestion, `reject` at TRIAL goes straight to FAILED, no transient REJECTED state is required, and `reboot` has no effect on TRIAL state.

Components that can be overwritten at device runtime (while the Update Client is running), may not require TRIAL state either. A component like this could have the even simpler flow:

![image](https://user-images.githubusercontent.com/5690199/169880151-a5369f39-fd17-42c1-bcec-1f32bb2b8954.png)